x/crosschain/keeper/keeper_out_tx_tracker.go

```diff
@@ -271,6 +268,76 @@ func (k msgServer) AddToOutTxTracker(goCtx context.Context, msg *types.MsgAddToO
 	return &types.MsgAddToOutTxTrackerResponse{}, nil
 }
 
+// ValidateEVMOutTxBody validates the sender address, nonce and chain ID.
+// Note: 'msg' may contain fabricated information
+func ValidateEVMOutTxBody(msg *types.MsgAddToOutTxTracker, txBytes []byte, tssEth string) error {
+	var txx ethtypes.Transaction
+	err := txx.UnmarshalBinary(txBytes)
+	if err != nil {
+		return err
+	}
+	signer := ethtypes.NewLondonSigner(txx.ChainId())
+	sender, err := ethtypes.Sender(signer, &txx)
+	if err != nil {
+		return err
+	}
+	tssAddr := eth.HexToAddress(tssEth)
+	if tssAddr == (eth.Address{}) {
+		return fmt.Errorf("tss address not found")
+	}
+	if sender != tssAddr {
+		return fmt.Errorf("sender %s is not tss address", sender)
+	}
+	if txx.ChainId().Cmp(big.NewInt(msg.ChainId)) != 0 {
+		return fmt.Errorf("want evm chain id %d, got %d", txx.ChainId(), msg.ChainId)
+	}
+	if txx.Nonce() != msg.Nonce {
+		return fmt.Errorf("want nonce %d, got %d", txx.Nonce(), msg.Nonce)
+	}
+	if txx.Hash().Hex() != msg.TxHash {
+		return fmt.Errorf("want tx hash %s, got %s", txx.Hash().Hex(), msg.TxHash)
+	}
+	return nil
+}
+
+// ValidateBTCOutTxBody validates the SetWit sender address, nonce and chain ID.
+// Note: 'msg' may contain fabricated information
+func ValidateBTCOutTxBody(msg *types.MsgAddToOutTxTracker, txBytes []byte, tssBtc string) error {
+	tx, err := btcutil.NewTxFromBytes(txBytes)
+	if err != nil {
+		return err
+	}
+	for _, vin := range tx.MsgTx().TxIn {
+		if len(vin.Witness) != 2 { // outTx is SegWit transaction for now
+			return fmt.Errorf("not a SegWit transaction")
+		}
+		pubKey, err := btcec.ParsePubKey(vin.Witness[1], btcec.S256())
+		if err != nil {
+			return fmt.Errorf("failed to parse public key")
+		}
+		addrP2WPKH, err := btcutil.NewAddressWitnessPubKeyHash(btcutil.Hash160(pubKey.SerializeCompressed()), config.BitconNetParams)
+		if err != nil {
+			return fmt.Errorf("failed to create P2WPKH address")
+		}
+		if addrP2WPKH.EncodeAddress() != tssBtc {
+			return fmt.Errorf("sender %s is not tss address", addrP2WPKH.EncodeAddress())
+		}
+	}
+	if common.BtcChainID() != msg.ChainId {
+		return fmt.Errorf("want btc chain id %d, got %d", common.BtcChainID(), msg.ChainId)
+	}
+	if len(tx.MsgTx().TxOut) < 1 {
+		return fmt.Errorf("outTx should have at least one output")
+	}
+	if tx.MsgTx().TxOut[0].Value != common.NonceMarkAmount(msg.Nonce) {
+		return fmt.Errorf("want nonce mark %d, got %d", tx.MsgTx().TxOut[0].Value, common.NonceMarkAmount(msg.Nonce))
+	}
+	if tx.MsgTx().TxHash().String() != msg.TxHash {
+		return fmt.Errorf("want tx hash %s, got %s", tx.MsgTx().TxHash(), msg.TxHash)
+	}
```

Comment:
Can we create an issue for the test implementation of those files?
Are `ValidateEVMOutTxBody` and `ValidateBTCOutTxBody`? It seems to me there could be made more generalized to be used with the in tx tracker as well

Reply:
That's a cool idea. As far as I can see, the `ChainID`, `TxHash`, `Nonce (evm only)` checking are general for both outTx and inTx. Other parameters' checking can be different (). I can reach out to @kingpinXD and see what checks he's going to perform.

Reply:
We can see during the merge